### PR TITLE
Introduce UI Element which supports an internal popup dialog

### LINKED ISSUE #468: Collected artifact info
In the original game artifact info can be seen with the [right click] over icon.
Our version has all information shown in a pop up window.
Artifact info

Artifact info2

### LINKED ISSUE #1726: Adventure map, empty pyramid info window, wrong horizontal position of bad luck sprites in window
In fheroes2, the horizontal position of bad luck sprites is not the same as in the original game.

In original game:

Pyramid luck empty sprite position H2 01

In fheroes2:

Pyramid luck empty sprite position fh2 01

### LINKED ISSUE #1969: Sea event operation
In the OG we don't hear any sounds, while event appears. Also we can see a message in the appearing window together with artifact.
image
In fheroes2 we first see a separate window with message.
image
And then we see the ordinary window while finding artifact. Also, we can hear "collecting" sound, while this window appears. In the OG we don't hear any sounds.
image

### LINKED ISSUE #4616: Game, Artifact graphic, Artifact graphic is not positioned correctly inside the frame
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

In fheroes2, the Artifact graphic is not positioned correctly inside the frame.
This is valid for all places where the Artifact graphic is displayed.
This flaw has been forgotten in closed issues: https://github.com/ihhub/fheroes2/issues/1366, https://github.com/ihhub/fheroes2/issues/1348.

In fheroes2 (in the example, in a Skeleton window):

Missing title skeleton window 04

### Save file

Not needed.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by ihhub**
- Kingdom Income window was changed to have `Kingdom Income` title and `Kingdom Income per day` to be more obvious for a player
- Events on land and sea now show artifact and resources in one window ( close #1969 )
- Collected artifacts, treasure chests, shrines related dialogs contains right mouse clickable icons ( close #468 )
- Pyramid visit outcome show correct horizontal position of icons and all icons are right mouse clickable ( close #1726 )
- Icons in captured mines and sawmills dialogs are right mouse clickable
- Experience related window such as in Daemon Cave and Gazebo have right mouse clickable icons
- Add static methods for Font types
- Fix Artifact related drawings ( close #4616 )
- Code cleanup and adding more assertions

### CODE REVIEW

File: src/fheroes2/gui/ui_dialog.h [hunk 1]
```diff
@@ -20,9 +20,136 @@
...
+        // This element must be cached to avoid heavy calculations.
+        Size _area;
+    };
+
+    class ArtifactDialogElement : public DialogElement
+    {
+    public:
+        explicit ArtifactDialogElement( const Artifact & artifact );
+
+        ~ArtifactDialogElement() override = default;
```
Comment (oleg-derevenetz) [inline src/fheroes2/gui/ui_dialog.h:65]: I believe this is not necessary to explicitly override a virtual destructor of the base class by the default destructor here and in the other similar places below, but it's a matter of taste :)

**Review by oleg-derevenetz — APPROVED** [review]
Hi @ihhub LGTM, just one optional comment.

File: src/fheroes2/gui/ui_dialog.h (hunk 1 above)
Comment (ihhub) [reply src/fheroes2/gui/ui_dialog.h:65]: I did this as clang-tidy as well as compilers were complaining about it.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , could you please take a look again? I expanded the changes to have a completed pull request. I removed a lot of old dialog based code and fixed some issues as well standardized Text types.

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hi @Branikolog and @LeHerosInconnu, I am asking you to help to test these changes as they impact a lot of dialogs.

I don't have too much time right now, I'll look at this towards the end of next week. :)

**Comment by zenseii** [comment]
This looks very good! I only found some small points that could be discussed or fixed:

Fixed, in my opinion:

1. Right-clickable experience icon and gold icon when picking up a treasure chest.

2. Experience info English text - Instead of "go up a level" I would say "gain a level" just to stay coherent with the rest of the game.

3. Experience info English text - Add text saying that gaining a level let's you learn a new secondary skill or improve one (yes, at high levels you won't gain anything new when maxed out)

Discussable:

1. Pyramid/Shrine/Etc. - In case you cannot learn a spell, should the spell info still be possible to be found by right-clicking something?

2. Right-clickable resource in the lower right status window when you pick up something on the adventure map?

**Review by felix642 — APPROVED** [review]

**Comment by ihhub** [comment]
> This looks very good! I only found some small points that could be discussed or fixed:
> 
> Fixed, in my opinion:
> 
>     1. Right-clickable experience icon and gold icon when picking up a treasure chest.
> 
>     2. Experience info English text - Instead of "go up a level" I would say "gain a level" just to stay coherent with the rest of the game.
> 
>     3. Experience info English text - Add text saying that gaining a level lets you learn a new secondary skill or improve one (yes, at high levels you won't gain/improve any new secondary skills when maxed out.)
> 
> 
> Discussable:
> 
>     1. Pyramid/Shrine/Etc. - In case you cannot learn a spell, because of insufficient Wisdom skill, should the spell info still be possible to find by right-clicking something? I believe there's an issue/discussion already concerning this.
> 
>     2. Right-clickable resource in the lower right status window when you pick up something on the adventure map?
> 
> 
> Thank you for the work put into this PR!

Hi @zenseii , thanks for your review!
Regarding your comments:
1. It is not a part of the standard dialog introduced in this pull request. It will be fixed later.
2 and 3. The text appears to be the same as in the original game. I wrote in the exact same way. We can refine it later.

Discussable notes:
1. It can be done with `CustomImageDialogElement` class but as you remarked it is debatable.
2. This window is not a part of the standard window classes.

**Comment by ihhub** [comment]
I am merging this pull request as it is to do not hold on any further. If we find something broken or needed an improvement we are going to fix it in a separate pull request.

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hello @ihhub,
> 
> > Hi @Branikolog and @LeHerosInconnu, I am asking you to help to test these changes as they impact a lot of dialogs.
> 
> I don't have too much time right now, I'll look at this towards the end of next week. :)

I spotted a few problems while testing this, I will report them in new issues. :)

**Pull request merged**